File: tools/layout.py
"""Parses the Canvas layout reply (the answer to `00` then `80`) and places squares on a grid.

The encoding is inferred from bench readings (docs/panel-bus.md, "Layout encoding"):
- `0xC0 | side`: a square. `side` is its own side facing its parent.
- After a square come its other three sides, clockwise from the entry side (seen from the front),
  each `0x00` for empty or the attached node's subtree.
- `0x90 | n` plus one byte: a power supply.
- `0x40` ends the stream; sides not listed yet are empty.

The drawing uses the root square's own frame, so the real wall may show it rotated in 90° steps.
"""

from __future__ import annotations

from dataclasses import dataclass

# Clockwise as seen from the front, matching the side numbering in square headers.
DOWN, LEFT, UP, RIGHT = range(4)
STEP = {DOWN: (0, -1), LEFT: (-1, 0), UP: (0, 1), RIGHT: (1, 0)}
EMPTY = 0x00
END = 0x40
SQUARE = 0xC0
PSU = 0x90


class LayoutError(ValueError):
    pass


@dataclass(frozen=True)
class Node:
    kind: str  # "square" or "psu"
    header: int
    x: int
    y: int
    index: int | None = None  # squares only: bus order, as used by F8 addressing


@dataclass(frozen=True)
class Layout:
    nodes: list[Node]
    controller_side: int

    @property
    def squares(self) -> list[Node]:
        return [node for node in self.nodes if node.kind == "square"]


class _Reader:
    def __init__(self, data: list[int]):
        self.data = data
        self.position = 0
        self.ended = False

    def next(self) -> int | None:
        """The next byte, or None once the end marker has been read."""
        if self.ended:
            return None
        if self.position >= len(self.data):
            raise LayoutError(f"layout reply ended without {END:02X}: {_hex(self.data)}")
        value = self.data[self.position]
        self.position += 1
        if value == END:
            self.ended = True
            return None
        return value


def parse_layout(reply: list[int]) -> Layout:
    reader = _Reader(reply)
    header = reader.next()
    if header is None or header & 0xF0 != SQUARE:
        raise LayoutError(f"expected a square header first, got {_hex(reply)}")
    controller_side = header & 0x0F
    nodes: list[Node] = []
    _read_square(reader, header, 0, 0, controller_side, nodes)
    if reader.position != len(reply):
        raise LayoutError(f"bytes left after the layout ended: {_hex(reply[reader.position :])}")

    occupied: dict[tuple[int, int], Node] = {}
    for node in nodes:
        other = occupied.setdefault((node.x, node.y), node)
        if other is not node:
            raise LayoutError(f"two nodes placed at ({node.x}, {node.y}): the encoding model is wrong for this layout")
    return Layout(nodes, controller_side)


def _read_square(reader: _Reader, header: int, x: int, y: int, entry_direction: int, nodes: list[Node]) -> None:
    square_count = sum(node.kind == "square" for node in nodes)
    nodes.append(Node("square", header, x, y, index=square_count))
    for turn in (1, 2, 3):
        direction = (entry_direction + turn) % 4
        value = reader.next()
        if value is None or value == EMPTY:
            continue
        dx, dy = STEP[direction]
        if value & 0xF0 == SQUARE:
            # The neighbour is entered from the side facing back towards this square.
            _read_square(reader, value, x + dx, y + dy, (direction + 2) % 4, nodes)
        elif value & 0xF0 == PSU:
            reader.next()  # trailing byte, 00 on every reading so far
            nodes.append(Node("psu", value, x + dx, y + dy))
        else:
            raise LayoutError(f"unknown node byte {value:02X} at offset {reader.position - 1} in {_hex(reader.data)}")


def _turn(x: int, y: int, quarter_turns: int) -> tuple[int, int]:
    """Rotates a grid position clockwise by 90° per quarter turn."""
    for _ in range(quarter_turns % 4):
        x, y = y, -x
    return x, y


def render(layout: Layout, quarter_turns: int = 0) -> str:
    """ASCII grid: [n] squares in bus order, PSU, and the Pico on the root's entry side.

    `quarter_turns` rotates the drawing clockwise, since the root square's frame need not match the wall.
    """
    cells = {
        _turn(node.x, node.y, quarter_turns): f"[{node.index:^3}]" if node.kind == "square" else " PSU "
        for node in layout.nodes
    }
    dx, dy = STEP[layout.controller_side]
    cells[_turn(dx, dy, quarter_turns)] = "Pico "
    xs = [x for x, _ in cells]
    ys = [y for _, y in cells]
    rows = []
    for y in range(max(ys), min(ys) - 1, -1):
        rows.append("".join(cells.get((x, y), "     ") for x in range(min(xs), max(xs) + 1)).rstrip())
    return "\n".join(rows)


def _hex(data: list[int]) -> str:
    return " ".join(f"{value:02X}" for value in data)

File: pico/src/layout.c
#include "layout.h"

#include <stdbool.h>

#define EMPTY 0x00
#define END 0x40
#define SQUARE 0xC0
#define PSU 0x90
#define READ_ENDED (-1)
#define READ_ERROR (-2)

// Clockwise as seen from the front: down, left, up, right.
static const int8_t STEP_X[4] = {0, -1, 0, 1};
static const int8_t STEP_Y[4] = {-1, 0, 1, 0};

typedef struct {
    const uint8_t *data;
    size_t length;
    size_t position;
    bool ended;
    layout_position_t *positions;
    size_t max;
    size_t count;
    bool failed;
} parser_t;

static int next_byte(parser_t *parser) {
    if (parser->ended) return READ_ENDED;
    if (parser->position >= parser->length) {
        parser->failed = true;  // no end marker
        return READ_ERROR;
    }
    uint8_t value = parser->data[parser->position++];
    if (value == END) {
        parser->ended = true;
        return READ_ENDED;
    }
    return value;
}

static void read_square(parser_t *parser, int x, int y, int entry_direction) {
    if (parser->count >= parser->max || x < INT8_MIN || x > INT8_MAX || y < INT8_MIN || y > INT8_MAX) {
        parser->failed = true;
        return;
    }
    for (size_t i = 0; i < parser->count; i++) {
        if (parser->positions[i].x == x && parser->positions[i].y == y) {
            parser->failed = true;  // the model placed two squares in one spot
            return;
        }
    }
    parser->positions[parser->count++] = (layout_position_t){(int8_t)x, (int8_t)y};

    for (int turn = 1; turn <= 3 && !parser->failed; turn++) {
        int direction = (entry_direction + turn) % 4;
        int value = next_byte(parser);
        if (value == READ_ERROR) return;
        if (value == READ_ENDED || value == EMPTY) continue;
        if ((value & 0xF0) == SQUARE) {
            // The neighbour is entered from the side facing back towards this square.
            read_square(parser, x + STEP_X[direction], y + STEP_Y[direction], (direction + 2) % 4);
        } else if ((value & 0xF0) == PSU) {
            if (next_byte(parser) == READ_ERROR) return;  // trailing byte, 00 on every reading so far
        } else {
            parser->failed = true;
        }
    }
}

int layout_positions(const uint8_t *reply, size_t length, layout_position_t *positions, size_t max) {
    parser_t parser = {.data = reply, .length = length, .positions = positions, .max = max};
    int header = next_byte(&parser);
    if (header < 0 || (header & 0xF0) != SQUARE) return -1;
    read_square(&parser, 0, 0, header & 0x0F);
    if (parser.failed || parser.position != length) return -1;
    return (int)parser.count;
}
